### PR TITLE
Add marko support to storybooksJS

### PR DESCRIPTION
**Pull request by nm123github**
## What I did
Marko support for storybooksJS.

## How to test

Is this testable with Jest or Chromatic screenshots?
Not sure.

Does this need a new example in the kitchen sink apps?
Yes, i have added marko-cli in examples folder.

Does this need an update to the documentation?
It does. I have added some preliminary docs (basically just copied from other apps and edited/added some basic stuff). Will keep adding as we go.

Do let me know what might be missing. I still had a quick question around arranging my stories in marko-cli, and once i get that answered i might change up some stuff. Also, couldn't add any tags because i don't think i have permissions for the same.

### CODE REVIEW

**Comment by wuweiweiwu** [comment]
related: https://github.com/storybooks/storybook/issues/2841

**Review by wuweiweiwu — CHANGES_REQUESTED** [review]
Check list:

- make sure yarn lint passes
- add test for marko knobs addon
- add [config.js](https://github.com/storybooks/storybook/blob/master/examples/official-storybook/config.js#L11) for hierarchal story separation. (right now it displays the separator `'|'`)

File: postcss.config.js [hunk 1]
```diff
@@ -0,0 +1 @@
+module.exports = {};
```
Comment (wuweiweiwu) [inline postcss.config.js]: I don't think we need a config file for postcss if it is just empty.

**Review by wuweiweiwu — CHANGES_REQUESTED** [review]

**Comment by nm123github** [comment]
Working on lint erros 👍

File: postcss.config.js (hunk 1 above)
Comment (nm123github) [reply postcss.config.js]: Was facing this [No PostCSS Config found](https://github.com/akveo/ngx-admin/issues/604) issue. Solution seemed to be adding an empty postcss.config.js. I will look into if maybe i can resolve this some other way.

File: app/marko/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+For more information visit: [storybook.js.org](https://storybook.js.org)
+
+* * *
+
+Storybook also comes with a lot of [addons](https://storybook.js.org/addons/introduction) and a great API to customize as you wish.
+You can also build a [static version](https://storybook.js.org/basics/exporting-storybook) of your storybook and deploy it anywhere you want.
+
+Here are some featured storybooks that you can reference to see how Storybook works:
+
+## Typescript
```
Comment (igor-dv) [inline app/marko/README.md]: Did you want to add here something?

File: app/marko/demo.js [hunk 3]
```diff
@@ -0,0 +1 @@
+module.exports = {};
```
Comment (igor-dv) [inline app/marko/demo.js]: Is this file needed?

File: app/marko/package.json [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{
+  "name": "@storybook/marko",
+  "version": "4.0.0-alpha.1",
```
Comment (igor-dv) [inline app/marko/package.json]: change the version to the latest

File: app/marko/package.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    "build-storybook": "./bin/build.js",
+    "start-storybook": "./bin/index.js",
+    "storybook-server": "./bin/index.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/storybooks/storybook.git"
+  },
+  "scripts": {
+    "dev": "cross-env DEV_BUILD=1 nodemon --watch ./src --exec \"yarn prepare\"",
```
Comment (igor-dv) [inline app/marko/package.json]: this is already unneeded.

File: app/marko/package.json [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    "html-loader": "^0.5.5",
+    "html-webpack-plugin": "^3.2.0",
+    "lodash.flattendeep": "^4.4.0",
+    "markdown-loader": "^2.0.2",
+    "prop-types": "^15.6.1",
+    "webpack": "^4.5.0",
+    "webpack-hot-middleware": "^2.21.2"
+  },
+  "devDependencies": {
+    "nodemon": "^1.17.3"
```
Comment (igor-dv) [inline app/marko/package.json]: this is also unneeded

File: app/marko/src/client/preview/render.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
+/* global document */
```
Comment (igor-dv) [inline app/marko/src/client/preview/render.js]: After the latest changes @Hypnosphi did, most of this code is redundant (moved to core) as well as "NoPreview.marko" and "ErrorDisplay.marko". take a look at the other apps to find the change.

File: app/marko/src/server/config/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    plugins: [
+      new HtmlWebpackPlugin({
+        filename: 'index.html',
+        chunks: ['manager'],
+        chunksSortMode: 'none',
+        data: {
+          managerHead: getManagerHeadHtml(configDir),
+          version,
+        },
+        template: require.resolve('../index.html.ejs'),
```
Comment (igor-dv) [inline app/marko/src/server/config/webpack.config.js]: some parts of this file are also moved to core (indexHtmlPath, iframeHtmlPath)

File: app/marko/src/server/config/webpack.config.prod.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    plugins: [
+      new HtmlWebpackPlugin({
+        filename: 'index.html',
+        chunks: ['manager', 'runtime~manager'],
+        chunksSortMode: 'none',
+        data: {
+          managerHead: getManagerHeadHtml(configDir),
+          version,
+        },
+        template: require.resolve('../index.html.ejs'),
```
Comment (igor-dv) [inline app/marko/src/server/config/webpack.config.prod.js]: also here

File: app/marko/src/server/iframe.html.ejs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
```
Comment (igor-dv) [inline app/marko/src/server/iframe.html.ejs]: this was moved to core

File: app/marko/src/server/index.html.ejs [hunk 11]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE html>
```
Comment (igor-dv) [inline app/marko/src/server/index.html.ejs]: and this =)

File: examples/marko-cli/.editorconfig [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# top-most EditorConfig file
+root = true
```
Comment (igor-dv) [inline examples/marko-cli/.editorconfig]: probably this is not a root, since we already have a root =)

File: examples/marko-cli/package.json [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  "license": "MIT",
+  "devDependencies": {
+    "@storybook/marko": "4.0.0-alpha.1",
+    "@storybook/addon-actions": "4.0.0-alpha.4",
+    "@storybook/addon-backgrounds": "4.0.0-alpha.4",
+    "@storybook/addon-centered": "4.0.0-alpha.4",
+    "@storybook/addon-knobs": "4.0.0-alpha.4",
+    "@storybook/addon-links": "4.0.0-alpha.4",
+    "@storybook/addon-notes": "4.0.0-alpha.4",
+    "@storybook/addon-options": "4.0.0-alpha.4",
```
Comment (igor-dv) [inline examples/marko-cli/package.json]: I see these addons here. Do all of them supported and do they all of them have examples here?

File: examples/marko-cli/src/components/welcome/index.marko [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
...ho7v649m7vvXW6s3l5GNfs2/zRdroMTZbwnPsz9e03pjrSS9UjDhIFDIxiezwRehu51w1Zu6z7TjzIuqXVYNOLj33hQll4eN0JQmDUOqRdpZ5YIuEk8rm3mjISW+sE1jKKo9hLZsPPcPXRXL3Xa5OJRNTCp5LTnpDigt8L6VpkuO7D1+6rv1/CS6by8lH6ku1ENKBqtaVanmgT7nQ7o6Cly+2ODl8G8uyQvyvEh+bNulQRdnDaZAmXZ2143nP7A+Tx15zJJSsNvCE1rhwrnORuaPk5S5lOPYY6kCCdKiMuEy92Ty48DdMW65zHrRPEM5xg0nLt9/m61I+f2o3jK6AXg/lRobdBzrZkpLx3HXFdNqCvnlovdEEGcfNHGkDFNC2sCCUWHbHUYs8x7kX5lsJPg6YDcRwCFvENIU6jG5qsipdhX7gKzfS3aYB/VyYW6Imoal/mW762wI3kU62oDKJ0gxTdDSTTj4qf9KXd91S7iunlj/X701k4AEufFzCWFIzjCtNlZRHAJdSCBM8eFtBcHkpM8bconvbzis+WhxmbzSbGLbP3dr8kTa6bJW7/e6bn8xjucxqGGPSQF2AvlB4hHfmDSldi0j3K+/1/i3uifxqz4l7Iq1WJkieJMuah+/8prFjv77P5c+eTV2LDEqVB7iclKlQP7NQ+UdwOs908iTAYCpYIxj0cEW0G6exmeraJsQw2lXS6Sxfs/ARn2YEiMRuu5zfmEgcAG800PmTgLox/UMW0VwpoSpqZyb9Ri6d5QahwuyNhsZloIEP4kALh7oc2aUGO2HqRc+cUXC8H0G3yca4E287dDBAg+kEhOm7bPnfM84x7KfHruL9IHHlFzwAYdOw+j2/Z1sZpvJ7TXW/AiqbVEZCyQJAuz1NlzTbS/xIn5E+V/Qv6DATlU2zfQVje8uTJv6Gado2qIPgMuD5L+oY6LtFeUorn6F2iHuV+eK68l7vayrDn8e6lQ3Dc3o6/23n1esf4+wwa/OxLyJgUS39DDCrFdiM8qivmAN3a6x56wOA9uD+VIZSNvxyhDFA61yJaaZr6IvZ/Ud/iK6hHxHSFkjNf5VNcLPsYIy4jaUczjGumJcZmy1qPIa1i/5V3BrQpUB+yxibrCpsF9dlxOQPLVzw7URcgp7JjL2KTgTWoBkQ0Oq00P8HD2rZq0DBdtAAAAAASUVORK5CYII=');
```
Comment (igor-dv) [inline examples/marko-cli/src/components/welcome/index.marko]: OMG 😱

**Review by igor-dv — COMMENTED** [review]
Very, very cool 🔥!

Missing things ( I understand that this is a WIP though =) ):
1. addons support in the ADDONS_SUPPORT.md
2. Live examples (netlify)
3, Probably also the "CLI" integration

File: examples/marko-cli/src/components/welcome/index.marko (hunk 14 above)
Comment (nm123github) [reply examples/marko-cli/src/components/welcome/index.marko]: I know ☺ . i still need to figure out how to load relative images correctly.

File: postcss.config.js (hunk 1 above)
Comment (nm123github) [reply postcss.config.js]: Added (empty) options inline in js (at /storybook/lib/core/src/server/config/defaults/webpack.config.js) where the postcss-loader is being applied.

File: examples/marko-cli/package.json (hunk 13 above)
Comment (nm123github) [reply examples/marko-cli/package.json]: Not yet. But plan is to add support soon!

File: app/marko/package.json (hunk 5 above)
Comment (nm123github) [reply app/marko/package.json]: you mean dev command?

**Comment by Hypnosphi** [comment]
Please integrate the changes from https://github.com/storybooks/storybook/pull/3472

File: app/marko/package.json (hunk 4 above)
Comment (nm123github) [reply app/marko/package.json]: Some reason when i do that i get below error when i do a yarn bootstrap

> error An unexpected error occurred: "https://registry.yarnpkg.com/@storybook%2fmarko: Not found".

**Comment by nm123github** [comment]
Hello all, I think i have taken care of most issues. After the new refactor the render.js has indeed become a lot simpler.

Few open items i can think of which i might need some help with
- Should my package.json version for @storybook/marko be 4.0.0-alpha.4? If i change it to that and do a yarn bootstrap i get some error
> error An unexpected error occurred: "https://registry.yarnpkg.com/@storybook%2fmarko: Not found".
- Live examples (netlify) **?**
- Probably also the "CLI" integration **?**
- Don't see any examples for tests in other apps?
- Add config.js for hierarchal story separation. (right now it displays the separator '|')?  What exactly is this, and why do we need it?

File: app/marko/package.json (hunk 4 above)
Comment (Hypnosphi) [reply app/marko/package.json]: You probably need to change it in example dependency as well

**Comment by nm123github** [comment]
Hey guys, i think i have got it to a point where i knew what needs to fixed. Let me know if we're good to go. Feel free to slack me (or #markojs) if thats quicker.

**Comment by Hypnosphi** [comment]
Plus https://github.com/storybooks/storybook/pull/3520 please

**Comment by Hypnosphi** [comment]
[Live Example](https://deploy-preview-3504--storybooks-marko.netlify.com/)

@nm123github "no preview" story shows an error https://deploy-preview-3504--storybooks-marko.netlify.com/?selectedKind=Hello&selectedStory=with%20No%20Preview%21&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

File: app/marko/src/demo/Button.marko [hunk 15]
```diff
@@ -0,0 +1,12 @@
+class {
+    onCreate(input) {
+    }
```
Comment (Hypnosphi) [inline app/marko/src/demo/Button.marko]: Why is this empty method needed?

File: app/marko/src/server/config/globals.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+/* globals window */
```
Comment (Hypnosphi) [inline app/marko/src/server/config/globals.js]: please use `import { window } from 'global'` instead

File: app/marko/src/server/config/webpack.config.js [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+      new webpack.HotModuleReplacementPlugin(),
+      new CaseSensitivePathsPlugin(),
+      quiet ? null : new webpack.ProgressPlugin(),
+      new Dotenv({ silent: true }),
+    ].filter(Boolean),
+    module: {
+      rules: [
+        {
+          test: /\.marko$/,
+          loader: require.resolve('marko-loader'),
```
Comment (Hypnosphi) [inline app/marko/src/server/config/webpack.config.js]: you need to add it as a direct dependency of `@storybook/marko`

Comment (nm123github) [reply app/marko/src/server/config/webpack.config.js]: Should we remove it from the root package.json?

**Comment by nm123github** [comment]
Also, thanks @Hypnosphi for taking care of some of the pending items.

**Comment by igor-dv** [comment]
> Let me know if we're good to go

Regarding the addons, if you think you won't get to them in this PR, let's remove the unneeded deps in the example. Also, configuring Storyosurce is easy, so probably worth adding it here.

File: app/marko/src/server/config/webpack.config.js (hunk 17 above)
Comment (Hypnosphi) [reply app/marko/src/server/config/webpack.config.js]: As you wish

Comment (nm123github) [reply app/marko/src/server/config/webpack.config.js]: ok. will just leave it there.

**Review by Hypnosphi — APPROVED** [review]

**Review by wuweiweiwu — APPROVED** [review]
Nice! 👍

**Comment by ndelangen** [comment]
Congrats @storybooks/team @nm123github Great work!

**Comment by Hypnosphi** [comment]
released as [`4.0.0-alpha.6`](https://github.com/storybooks/storybook/releases/tag/v4.0.0-alpha.6)
